pkg/source/hdfsprotocol/hdfs_source_client.go

```diff
@@ -17,52 +17,232 @@
 package hdfsprotocol
 
 import (
+	"bytes"
 	"context"
 	"io"
+	"io/ioutil"
+	"net/url"
+	"os/user"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"d7y.io/dragonfly/v2/cdnsystem/daemon/cdn"
+	"github.com/go-http-utils/headers"
+	"github.com/pkg/errors"
 
 	"d7y.io/dragonfly/v2/pkg/source"
+	"github.com/colinmarc/hdfs/v2"
 )
 
 const (
 	HDFSClient = "hdfs"
 )
+const (
+	layout = "2006-01-02 15:04:05"
+)
 
 func init() {
 	source.Register(HDFSClient, NewHDFSSourceClient())
 }
 
 // hdfsSourceClient is an implementation of the interface of SourceClient.
 type hdfsSourceClient struct {
+	sync.RWMutex
+	clientMap map[string]*hdfs.Client
 }
 
 type HDFSSourceClientOption func(p *hdfsSourceClient)
 
-func (h hdfsSourceClient) GetContentLength(ctx context.Context, url string, header source.RequestHeader) (int64, error) {
-	panic("implement me")
+func (h *hdfsSourceClient) GetContentLength(ctx context.Context, url string, header source.RequestHeader) (int64, error) {
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return -1, err
+	}
+	info, err := hdfsClient.Stat(path)
+	if err != nil {
+		return -1, err
+	}
+	return info.Size(), nil
+}
+
+func (h *hdfsSourceClient) IsSupportRange(ctx context.Context, url string, header source.RequestHeader) (bool, error) {
+	return true, nil
+}
+
+func (h *hdfsSourceClient) IsExpired(ctx context.Context, url string, header source.RequestHeader, expireInfo map[string]string) (bool, error) {
+	lastModified := expireInfo[headers.LastModified]
+	//eTag := expireInfo[headers.ETag]
+	if lastModified == "" {
+		return true, nil
+	}
+
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return false, err
+	}
+
+	info, err := hdfsClient.Stat(path)
+	if err != nil {
+		return false, err
+	}
+	t, err := time.ParseInLocation(layout, lastModified, time.Local)
+	if err != nil {
+		return false, err
+	}
+	return info.ModTime().Format(layout) != t.Format(layout), nil
+}
+
+func (h *hdfsSourceClient) Download(ctx context.Context, url string, header source.RequestHeader) (io.ReadCloser, error) {
+	hdfsClient, path, err := h.getHDFSClientAndPath(url)
+	if err != nil {
+		return nil, err
+	}
+	hdfsFile, err := hdfsClient.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer hdfsFile.Close()
+
+	data, err := ioutil.ReadAll(hdfsFile)
```

Comment:
Should not read all data to memory, just return the io.ReadCloser, the caller will close this.

Reply:
done